[WIP]: Factory contract

nagxsan opened the pull request.

augustin-v commented:
Thanks for the PR @nagxsan, I like the energy! I'll review it in a few hours

nagxsan commented:
No no this is far from ready, I just raised it to keep track of my progress 😅

augustin-v commented on contracts/src/contract/interface.cairo:
```diff
@@ -108,3 +109,14 @@ pub trait IERC20Trait<TContractState> {
 }
+
+#[starknet::interface]
+pub trait UniswapV3PoolDeployer<TContractState> {
```
Change name to `IUniswapV3PoolDeployer` for clarity

nagxsan replied:
Done

augustin-v commented on contracts/src/contract/univ3factory.cairo:
```diff
@@ -0,0 +1,56 @@
+    struct Storage {
+        parameters: PoolParameters,
+        tick_spacings: Map<i32, bool>,
+        pools: Map<ContractAddress, Map<ContractAddress, Map<i32, ContractAddress>>>,
```
Unisawp v3 factory contract is supposed to have an owner determined by token holders, but as we are just an educational project, for now you can just add the owner value to the storage and in the constructor do:
```rust
self.owner.write(0x08.try_into().unwrap()); // placeholder address
```

```suggestion
        pools: Map<ContractAddress, Map<ContractAddress, Map<i32, ContractAddress>>>,
        owner: ContratAddress
```

nagxsan replied:
Done

augustin-v commented on contracts/src/contract/univ3factory.cairo:
```diff
@@ -0,0 +1,56 @@
+    #[constructor]
+    fn constructor(ref self: ContractState) {
+        self.tick_spacings.write(10, true);
+        self.tick_spacings.write(60, true);
```
Nice job following the book here, I doubled checked and actually uniswap v3 also supports tick spacing of 200. [You can check here if interested, page 3 under the Governance thing](https://app.uniswap.org/whitepaper-v3.pdf). 
```suggestion
        self.tick_spacings.write(60, true);
        self.tick_spacing.write(200, true)
```

nagxsan replied:
Done

augustin-v commented on contracts/src/contract/univ3factory.cairo:
```diff
@@ -0,0 +1,56 @@
+
+        fn create_pool(
+            ref self: ContractState,
+            tokenX: ContractAddress,
+            tokenY: ContractAddress,
```
Rename to :
```suggestion
            token0: ContractAddress,
            token1: ContractAddress,
```

nagxsan replied:
Done

augustin-v reviewed (changes requested):
Looking promising impressive considering you just started cairo! The next step would be creating a pool with the 2 tokens  passed in the `create_pool` function. To do that we need to use starknet's [system calls](https://book.cairo-lang.org/appendix-08-system-calls.html#deploy), I recommend looking at the [YAS code](https://github.com/lambdaclass/yet-another-swap/blob/f3ee03a3564a37698e9589f564ac63aa59dab283/crates/yas_core/src/contracts/yas_factory.cairo#L188-L199) for implementation reference.

Also make sure to add eent. e.g: PoolCreated.

I'll be happy to help if you need :)

nagxsan commented:
Hey @augustin-v ! Thanks for the review and suggestions. I have added the implementation of the `create_pool` function and also added a `get_pool` function to retrieve address of the deployed pool. There are two things though I am a bit confused here:
- Is this `let pool_class_hash = get_class_hash_at_syscall(get_contract_address()).unwrap();` correct in `factory.cairo` at line 102? Or do we get the class hash from the user when the factory contract is being first deployed?
- In the `test_factory_create_pool` test, I am getting this error: `('Input too long for arguments')`. I am trying to understand but it is just `ContractAddress` and an `i32` value for `tick_spacing` so what is wrong here.

Thanks for the help!

augustin-v commented:
Great job @nagxsan ! The `input too long for arguments` error happens because the pool contract constructor expects 8 arguments and right now we only have 3. Let's add the pool deployment parameters to the PoolParameters struct and have it stored inside of our Storage. We need to add the 
```rust
tick_spacing: i32, // <- right now not supported by the pool contract, do not add add it yet when calling the pool contract constructor 
sqrt_pricex96: u256,
tick: i32,
tick_address: ContractAddress,
bitmap_address: ContractAddress,
position_address: ContractAddress
```
variables. Then we can update the factory to set these parameters before deployment, deploy with empty calldata, and have the pool constructor fetch the parameters from the factory via a helper get function

nagxsan commented:
Apologies for the delayed response. I am on it, will raise the changes in a day.

augustin-v commented:
> Apologies for the delayed response. I am on it, will raise the changes in a day.

no worries its a pretty hard component to implmement take your time, you'll be a chad cairo dev if you manage to implement it

nagxsan commented:
Thank you so much for your patience!
